frontend/src/lib/utils.tsx

```diff
@@ -442,18 +471,54 @@ export function copyToClipboard(value, description) {
     }
 }
 
-export function clamp(value, min, max) {
+export function clamp(value: number, min: number, max: number): number {
     return value > max ? max : value < min ? min : value
 }
 
-export function isMobile() {
+export function isMobile(): boolean {
     return navigator.userAgent.includes('Mobile')
 }
 
-export function isMac() {
+export function isMac(): boolean {
     return navigator.platform.includes('Mac')
 }
 
-export function platformCommandControlKey() {
+export function platformCommandControlKey(): string {
     return isMac() ? '⌘' : 'Ctrl'
 }
+
+export function groupBy<T>(items: T[], groupResolver: (item: T) => string | number): Record<string | number, T[]> {
+    const itemsGrouped: Record<string | number, T[]> = {}
+    for (const item of items) {
+        const group = groupResolver(item)
+        if (!(group in itemsGrouped)) itemsGrouped[group] = [] // Ensure there's an array to push to
+        itemsGrouped[group].push({ ...item })
```

Comment:
Nice touch with making a copy here... though I think it's actually not needed. At least it feels unintuitive that a simple operation to just group existing elements by keys also creates copies of those elements. We should probably just push the immutability down the chain and make sure no "client code" modifies these elements. It's also a significant performance hit if you just want to e.g. count the number of groups or something like that.

The uniqueBy below for example doesn't modify the items and I think it pays to be consistent.